Sync/BatchingClient.swift

```diff
@@ -0,0 +1,221 @@
+
+import Foundation
+import Alamofire
+import Shared
+import XCGLogger
+import Deferred
+
+// MARK: Batch Client Errors
+public class SerializeRecordFailure<T: CleartextPayloadJSON>: MaybeErrorType {
+    public let record: Record<T>
+
+    public var description: String {
+        return "Failed to serialize record: \(record)"
+    }
+
+    public init(record: Record<T>) {
+        self.record = record
+    }
+}
+
+private let log = Logger.syncLogger
+
+// MARK: Internal Types
+private typealias UploadRecord = (guid: GUID, payload: String, size: Int)
+
+/**
+ *  A payload represents a block of records sent in a single POST to the server within a batch.
+ *  The payload tries to contains as many records it can fit while respecting the limits imposed
+ *  by the info/configuration endpoint.
+ */
+private struct Payload {
+    let config: InfoConfiguration
+    private(set) var records: [UploadRecord] = []
+    private var bytes: ByteCount = 0
+    private var count: Int = 0
+
+    init(config: InfoConfiguration) {
+        self.config = config
+    }
+
+    /** Returns true if the record was added to this payload; false otherwise. */
+    mutating func add(record: UploadRecord) -> Bool {
+        if (record.size + bytes + 1) > config.maxPostBytes {
+            return false
+        }
+
+        if count >= config.maxPostRecords {
+            return false
+        }
+
+        records.append(record)
+        count += 1
+        bytes += record.size + 1     // For newline.
+        return true
+    }
+}
+
+/**
+ * Used to keep track of number of bytes and records sent as part of a batch
+ */
+private struct BatchMeta {
+    let config: InfoConfiguration
+    private var bytes: ByteCount = 0
+    private var count: Int = 0
+
+    init(config: InfoConfiguration) {
+        self.config = config
+    }
+
+    mutating func add(payload: Payload) -> Bool {
+        guard payloadFitsInBatch(payload) else {
+            return false
+        }
+
+        count += payload.count
+        bytes += payload.bytes
+        return true
+    }
+
+    private func payloadFitsInBatch(payload: Payload) -> Bool {
+        return (bytes + payload.bytes <= config.maxTotalBytes) && (count + payload.count < config.maxTotalRecords)
+    }
+}
+
+typealias BatchUploadFunction = (lines: [String], ifUnmodifiedSince: Timestamp?, queryParams: [NSURLQueryItem]?) -> Deferred<Maybe<StorageResponse<POSTResult>>>
+
+public class Sync15BatchClient<T: CleartextPayloadJSON> {
+    private let config: InfoConfiguration
+    private let uploader: BatchUploadFunction
+    private let serializeRecord: (Record<T>) -> String?
+    private var ifUnmodifiedSince: Timestamp?
+
+    private var batchMeta: BatchMeta
+    private var batchToken: BatchToken?
+    private var currentPayload: Payload
+    private var onCollectionUploaded: (POSTResult -> Void)?
+
+    // TODO: use I-U-S.
+    // Each time we do the storage operation, we might receive a backoff notification.
```

Comment:
Right now the backoff check is done in the `post` call from the collection client that gets passed in as the uploader function. The post call already handles filling a deferred with the `ServerInBackoffError` error. Tracing the `uploader` calls in the `BatchingClient` shows that this error will propagate up to the `addRecord` and `endBatch`. Since the `addRecord` would return a failure, external clients and bubble this up further following the same logic flow the existing `post` call has. I don't think any changes need to be done here.